cuML/src/common/cumlHandle.cpp

```diff
@@ -0,0 +1,418 @@
+{
+    destroyResources();
+    _dev_ids = dev_ids;
+    createResources();
+}
+
+int cumlHandle_impl::getDevice( int dev_idx ) const
+{
+    return _dev_ids[dev_idx];
+}
+
+int cumlHandle_impl::getNumDevices() const
+{
+    return _dev_ids.size();
+}
+
+void cumlHandle_impl::setStream( cudaStream_t stream )
+{
+    _userStream = stream;
+}
+
+cudaStream_t cumlHandle_impl::getStream() const
+{
+    return _userStream;
+}
+
+cudaStream_t cumlHandle_impl::getDeviceStream( int dev_idx ) const
+{
+    return _streams[dev_idx];
+}
+
+void cumlHandle_impl::setDeviceAllocator( std::shared_ptr<deviceAllocator> allocator )
+{
+    _deviceAllocator = allocator;
+}
+
+std::shared_ptr<deviceAllocator> cumlHandle_impl::getDeviceAllocator() const
+{
+    return _deviceAllocator;
+}
+
+void cumlHandle_impl::setHostAllocator( std::shared_ptr<hostAllocator> allocator )
+{
+    _hostAllocator = allocator;
+}
+
+std::shared_ptr<hostAllocator> cumlHandle_impl::getHostAllocator() const
+{
+    return _hostAllocator;
+}
+
+cublasHandle_t cumlHandle_impl::getCublasHandle( int dev_idx ) const
+{
+    return _cublas_handles[dev_idx];
+}
+
+cusolverDnHandle_t cumlHandle_impl::getcusolverDnHandle( int dev_idx ) const
+{
+    return _cusolverDn_handles[dev_idx];
+}
+
+cusparseHandle_t cumlHandle_impl::getcusparseHandle( int dev_idx ) const
+{
+    return _cusparse_handles[dev_idx];
+}
+
+void cumlHandle_impl::waitOnUserStream() const
+{
+    CUDA_CHECK( cudaEventRecord( _event, _userStream ) );
+    for (auto s : _streams)
+    {
+        CUDA_CHECK( cudaStreamWaitEvent( s, _event, 0 ) );
+    }
+}
+
+void cumlHandle_impl::waitOnInternalStreams() const
+{
+    for (auto s : _streams) 
+    {
+        CUDA_CHECK( cudaEventRecord( _event, s ) );
+        CUDA_CHECK( cudaStreamWaitEvent( _userStream, _event, 0 ) );
+    }
+}
+
+void cumlHandle_impl::createResources()
+{
+    CUDA_CHECK( cudaEventCreateWithFlags( &_event, cudaEventDisableTiming ) );
+    cudaDeviceScope _( 0 );
+    for (auto dev_id: _dev_ids) 
+    {
+        CUDA_CHECK( cudaSetDevice(dev_id) );
+
+        cudaStream_t stream;
+        CUDA_CHECK( cudaStreamCreate(&stream) );
+
+        cublasHandle_t cublas_handle;
+        CUBLAS_CHECK( cublasCreate(&cublas_handle) );
+        CUBLAS_CHECK( cublasSetStream(cublas_handle, stream) );
+
+        cusolverDnHandle_t cusolverDn_handle;
+        CUSOLVER_CHECK( cusolverDnCreate(&cusolverDn_handle) );
+        CUSOLVER_CHECK( cusolverDnSetStream(cusolverDn_handle, stream) );
+
+        cusparseHandle_t cusparse_handle;
+        CUSPARSE_CHECK( cusparseCreate(&cusparse_handle) );
+        CUSPARSE_CHECK( cusparseSetStream(cusparse_handle, stream) );
+
+        _streams.push_back(stream);
+        _cublas_handles.push_back( cublas_handle );
+        _cusolverDn_handles.push_back( cusolverDn_handle );
+        _cusparse_handles.push_back( cusparse_handle );
+    }
+}
+
+void cumlHandle_impl::destroyResources()
+{
+    while ( !_cusparse_handles.empty() )
```

Comment:
This is purely a aesthetic design critique, but a repeated pattern like this usually tells me there's a need for a more generic abstraction.

Perhaps a `for_each` over the elements of each vector and then `vector.clear()`.

Reply:
I agree to your statement. However given that I need to call a different destroy function for every vector I do not see how I would save a lot of code even using lambdas and `for_each`. However I am open for concrete suggestions :-)

Reply:
> However I am open for concrete suggestions :- )

Fair enough! 

Okay, so... this is a bit more of an invasive change, but hear me out because I think it's beneficial for more than just solving this problem.

So for your vectors of resources, instead of maintaining a vector of the "raw" resource, you could create a simple wrapper class for each resource. For example, streams:

```
struct stream_wrapper{

   stream_wrapper(){
      CUDA_CHECK( cudaStreamCreate(&stream) );
   }

   ~stream_wrapper(){
      CUDA_CHECK( cudaStreamDestroy(&stream) );
   }

   // This allows using a `stream_wrapper` implicitly as if it were a cudaStream_t*,
   // i.e., the fact that it's a wrapper is invisible to the user  
   operator cudaStream_t*() { return &stream }; 

private:
  cudaStream_t stream;
};
```

And then... your vector of `cudaStream_t` would become a vector of `stream_wrapper`s...

```
std::vector<stream_wrapper>           _streams;
```

And then in `destroyResources` you can simply do `_streams.clear()` and let the wrapper's destructor handle all of the resource specific cleanup! 

You could then create similar wrappers for all of the other resources.

What do you think?

Reply:
I see how this would make the code safer. However it will also add quite some code. Since this is an implementation detail and we are primarily discussing the API here I would leave the destroy resources code as it currently is. Any objections?

Reply:
> I see how this would make the code safer. However it will also add quite some code. Since this is an implementation detail and we are primarily discussing the API here I would leave the destroy resources code as it currently is. Any objections?

Absolutely just an implementation detail, don't let it block this PR.

Reply:
@jrhemstad can you file a issue on this one so that we don't miss this, please?